crypto/s2n_ecc_evp.c

```diff
@@ -129,14 +132,19 @@ static int s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_
     S2N_ERROR(S2N_ERR_ECDHE_GEN_KEY);
 }
 
-static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key, EVP_PKEY *peer_public, struct s2n_blob *shared_secret) {
+static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key, EVP_PKEY *peer_public, uint16_t iana_id, struct s2n_blob *shared_secret) {
     notnull_check(peer_public);
     notnull_check(own_key);
 
-    /* Peers MUST validate each other’s public key */
-    DEFER_CLEANUP(EC_KEY *ec_key = EVP_PKEY_get1_EC_KEY(peer_public), EC_KEY_free_pointer);
-    S2N_ERROR_IF(ec_key == NULL, S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
-    GUARD_OSSL(EC_KEY_check_key(ec_key), S2N_ERR_ECDHE_SHARED_SECRET);
+    /* From RFC 8446 Section 4.2.8.2: For the curves secp256r1 and secp384r1 peers MUST validate each other's
+     * public value Q by ensuring that the point is a valid point on the elliptic curve.
+     * For the curve x25519 the peer public-key validation check doesn't apply.
```

Comment:
Is there a reason you don't check != x25519? Which behavior should be the default?

Reply:
The reason I check for  ==secp256r1 and ==secp384r1 instead of !=x25519, is because the RFC states that only for these NIST curves (secp256r1 and secp384r1) peers MUST validate each other's  public value.